File: delight_oti/q4_Queue.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class Queue:
    def __init__(self):
        self.front = None
        self.back = None

    def peek(self):
        
        return self.front.data
    
    def enqueue(self, x):

        newNode = Node(x)
        
        if self.front is None:
            self.front = newNode
            self.back = newNode
        
        else:
            self.back.next = newNode
            self.back = newNode
    
    def dequeue(self):

        if self.front is None:
            return None
        else:
            x = self.front.data
            self.front = self.front.next

            if self.front is None:
                self.back = None
        
        return x
    
    def isEmpty(self):
        
        if self.front is None:
            return True
        else:
            return False
        
# 16mins

File: homework2/steven_law/q4_Queue.cpp
// Time spent: 22 mins 23 seconds

#include <iostream>
using namespace std;

struct Node {
    int data;
    Node* next;
};

class Queue {
private:
    Node* front;
    Node* back;

public:
    Queue() : front(nullptr), back(nullptr) {}

    // returns the first item in the queue. O(1) time.
    int peek() {
        return front->data;
    }

    // adds x to the back of the queue. O(1) time.
    void enqueue(int x) {

        

        Node* newNode = new Node();
        newNode->data = x;
        newNode->next = nullptr;

        if(back->next == nullptr){
            back = newNode;
            front = newNode;
            return;
        }

        back->next = newNode;
        back = newNode;

    }

    // removes and returns the first item in the queue. O(1) time.
    int dequeue() {
        if(front == nullptr) return -1;

        int first = front ->data;
        
        if(front->next == nullptr){
            delete front;
            front = nullptr;
            back = nullptr;
            return first;

        }
        Node* temp = front;
        front = front->next;
        delete temp; 

        return first;
    }

    // returns a boolean indicating whether the queue is empty. O(1) time.
    bool isEmpty() {
        return front == nullptr;
    }
};

int main() {

    return 0;
}
